### DISCUSSION TITLE
Adding Additional Rules via XML

### DISCUSSION THREAD

**Discussion by erik-buchholz** [ltex-plus/vscode-ltex-plus#96] [Q&A] [answered]
I was wondering if there is any way to add custom rules via XML to the bundled LanguageTool server?

For instance, I might have an XML rule discouraging the use of "it is":
```xml
<rule id="IT_IS" name="it is">
 <pattern>
  <token>it</token>
  <token>is</token>
 </pattern>
 <message>Try avoiding "it is" if possible.</message>
</rule>
```

This is possible when running a standalone LanguageTool server on localhost and adding the rules to `grammar_custom.xml`; however, I prefer the convenience of using the bundled LanguageTool server in LTeX.

**Comment by spitzerd** [comment, marked as answer]
The LanguageTool HTTP interface does not support custom rules, see https://languagetool.org/http-api/swagger-ui/#!/default/post_check
According to my knowledge, same is true for the Java interface.

You have to open an issue or pull request at the [LanguageTool repo](https://github.com/languagetool-org/languagetool) first. Afterwards, we can implement the feature into LTeX+

As you correctly wrote, you can use a standalone LanguageTool server instead of the bundled LanguageTool if you want to use custom rules.

**Comment by erik-buchholz** [comment]
Thanks for the quick response @spitzerd!

If anyone finds this topic in the future, here is a guide on how you can make LTeX(+) work with custom grammar rules.
This guide is for macOS, but will work similarly on Windows. You just have to change the paths.

## Use LanguageTool with Custom Rules

The following guide is based on [pl-lt-rules](https://github.com/JoeyEremondi/pl-lt-rules) but has beed adapted to work with the latest LanguageTool version (6.5).

### 1. Install LanguageTool

See: [Official Guide](https://dev.languagetool.org/http-server.html).

On macOS:
```bash
brew install languagetool
```

### 2. Get the path to the `grammar.xml` file

On macOS, the path is usually:
```bash
ls /opt/homebrew/Cellar/languagetool/*/libexec/org/languagetool/rules/en/grammar.xml
```

This path is different on Linux and Windows, make sure to adjust the path accordingly.

### 3. Set the environment variable `PATH_TO_GRAMMAR` to the output of the previous command

```bash
export PATH_TO_GRAMMAR=$(ls /opt/homebrew/Cellar/languagetool/*/libexec/org/languagetool/rules/en/grammar.xml)
```

Again, adjust the path if you are not on macOS.

### 4. Install the patch to the `grammar.xml` file

```bash
sed -e "s|home/username|$HOME|" lt.patch > lt.patch.tmp
patch -p1 $PATH_TO_GRAMMAR lt.patch.tmp
rm lt.patch.tmp
```

`lt.patch`: (Note that the patch has changed since the guide by Joey Eremondi)
```diff
--- /tmp/b	2022-06-06 16:37:42.000000000 -0700
+++ /tmp/a	2022-06-06 17:00:59.000000000 -0700
@@ -29,6 +29,7 @@
 -->

 <!DOCTYPE rules [
+    <!ENTITY UserRules SYSTEM "file:///home/username/.language-tool-rules.xml">
     <!ENTITY fractions "½|↉|⅓|⅔|¼|¾|⅕|⅖|⅗|⅘|⅙|⅚|⅐|⅛|⅜|⅝|⅞|⅑|⅒">
     <!ENTITY particles_of "de|von|van|of|d'|di|du|da|ibn|des|'s">
     <!ENTITY exceptions_multitoken_sensitive "(?-i)[A-Z]+\d+|[a-z][a-z]?\d+">
@@ -598,6 +599,10 @@
     </phrase>
   </phrases>

+  <category id="CUSTOM_USER_RULES" name="Custom User Rules" default="on">
+    &UserRules;
+  </category>
+
     <!--
     <category id="APRIL" name="April 1st">
         <rule id="APRIL_1ST" name="good advice">

```

You could change `file:///home/username/.language-tool-rules.xml` to the path of your custom rules so that you don't have to link the file below. In this case, you can skip steps 5-7. 

### 5. Define path to custom rules

```bash
export CUSTOM_RULES="PATH/TO/CUSTOM/RULES"
```

For example, you can use the rules from [pl-lt-rules](https://github.com/JoeyEremondi/pl-lt-rules). I might create a repository with my custom rules in the future, too. You can certainly find some inspiration how to write your own rules in the pl-lt-rules repository.

### 6. Link the custom rules to the specified location

```bash
ln -s $CUSTOM_RULES $HOME/.language-tool-rules.xml
```

### 7. Check that file exists

```bash
head -n 10 $HOME/.language-tool-rules.xml
```

### 8. [Optional] Define Path to N-gram model

```bash
export PATH_TO_NGRAM_MODEL=".../n-gram"
```

Replace the path. 
You can download the n-gram model from [here](https://languagetool.org/download/ngram-data/).
The directory structure should look like this:

```bash
.../n-gram
├── en
│   ├── 1grams
│   ├── 2grams
│   └── 3grams
```

### 9. [Optional] Define Path to LanguageTool Configuration

```bash
export PATH_TO_LT_CONFIG="/PATH/TO/LANGUAGETOOL/CONFIG"
```

You can see the available configuration options by running `languagetool-server --help`.

### 10. Start LanguageTool Server

```bash
# Define the configuration parameters (only use --languageModel if you have a n-gram model)
languagetool-server --languageModel $PATH_TO_NGRAM_MODEL --port 8081 --config $PATH_TO_LT_CONFIG --verbose
```

### 11. Use this server in the LTeX config (VSCode `settings.json`)

```json
{
    "ltex.languageToolHttpServerUri": "http://localhost:8081/",
}
```
